Add stylelint-scales

sixhours opened the pull request:
#### Changes proposed in this Pull Request

* Adds a plugin to stylelintrc that allows us to specify font-weights and other style properties to avoid straying from design best practices. See: https://www.npmjs.com/package/@signal-noise/stylelint-scales
* My first time adding something like this so please LMK if there's a process I need to follow or anything I should look out for. :)

#### Testing instructions

- Checkout this branch
- `npm ci`
- Try modifying a sass file and adding a bad font-weight value that isn't 400 or 700
- Try to commit locally
- stylelint should throw errrors on the precommit hook
<img width="1156" alt="Screen Shot 2020-04-10 at 1 45 50 PM" src="https://user-images.githubusercontent.com/1270189/79022187-366e4900-7b32-11ea-989f-3be417aa34e1.png">

- fixing the font-weight violations should allow for a local commit 

<img width="1418" alt="Screen Shot 2020-04-10 at 1 47 53 PM" src="https://user-images.githubusercontent.com/1270189/79022211-471ebf00-7b32-11ea-989b-4c5981541c6b.png">

sixhours committed:
Adds a style values plugin to stylelint to check for font-weights outside the required values.

gwwar commented:
Adding some testing instructions to summary 👍

gwwar reviewed (approved):
@sixhours nice, this tests well for me ✨ The addition looks correct to me too. As an aside, would we like to allow other rules like `inherit`?

simison commented:
Thanks! In an attempt to keep `npm run lint:css` as noise free as possible, could we run together (or in separate pr) with this change a conversion of keyed values to absolute values? So `normal` becomes 400 and bold becomes 700. Alternatively, whitelist those too. No big difference to me if we use absolute values or numbers.

There are quite a few: https://github.com/Automattic/wp-calypso/search?q=font-weight&type=

mmtr commented on package.json:
```diff
@@ -308,5 +308,8 @@
 		}
+	},
+	"devDependencies": {
+		"@signal-noise/stylelint-scales": "1.2.0"
```
I think our Calypso monorepo setup was preventing us from defining dependencies as `devDependencies` and we were always forced to add them as `dependencies`. Have this changed @Automattic/team-calypso ?

gwwar replied:
Good catch @mmtr, I didn't notice this was the only item in devDependencies! @Automattic/team-calypso is there some documentation on this that we could read up on?

gwwar replied:
🤔 Since stylelint is in dependencies, I guess we'd move it there? 

There's a broad overview in the following readme, but I didn't spot anything for build or dev deps.
https://github.com/Automattic/wp-calypso/blob/master/docs/monorepo.md

griffbrad replied:
Adding it to `dependencies` is the way to go, yes.  In the context of the root of the monorepo where you do your actual installs, there really isn't a practical difference between `dependencies` and `devDependencies`.  In that context, "dev" is really the overall purpose of the install.

kwight commented:
Awesome! I confirmed this worked with the dep moved to `dependencies` as suggested, so I think you're good to go @sixhours after that change. Side note, I'd also expect `inherit` to be a valid value, but up to you 👍

sixhours committed:
Move stylelint-scales to dependencies

sixhours commented:
> Thanks! In an attempt to keep `npm run lint:css` as noise free as possible, could we run together (or in separate pr) with this change a conversion of keyed values to absolute values? So `normal` becomes 400 and `bold` becomes 700. Alternatively, whitelist those too. No big difference to me if we use absolute or keyed values.

> Side note, I'd also expect inherit to be a valid value, but up to you 👍

Agreed on both points, but after trying it and doing some searching, it doesn't look like stylelint-scales supports non-numerical values like `inherit` or `bold`. :-/

Wondering if we could do an automated search/replace for font weights across Calypso, or if that has to be done manually... @gwwar what are your thoughts?

gwwar commented:
We could fork the plugin/file a patch upstream to include the values. https://github.com/signal-noise/stylelint-scales/blob/master/lib/rules/font-weight/index.js This doesn't look that complated

Otherwise, I'm not aware of popular ASTs built for SCSS (like we have for JS codemods). It'd be a search and replace over probably a number of PRs.
 
> Wondering if we could do an automated search/replace for font weights across Calypso, or if that has to be done manually... @gwwar what are your thoughts?

sixhours commented:
I'm inclined to deploy this as-is and work on a fork of the plugin to make it compatible with non-numerical font-weight values. That would mean some extra noise while we sort that fork out, though.

gwwar commented:
Please re-check package.lock :D We've just switched over to yarn, so please rebase and regenerate.

<img width="214" alt="Screen Shot 2020-04-16 at 10 09 01 AM 2" src="https://user-images.githubusercontent.com/1270189/79485585-538b9780-7fca-11ea-951b-68de2fed5ed7.png">

sixhours committed:
Remove unused package-lock.json, adds yarn.lock

sixhours commented:
> Please re-check package.lock :D We've just switched over to yarn, so please rebase and regenerate.

Whoops, now that rebase error I saw makes more sense. 😅 I think I've fixed it by removing `package-lock.json` and running `yarn install`, but let me know if there's something else I missed!

gwwar commented:
Yup the diff looks much more reasonable, thanks! Feel free to :shipit: if Circle is happy
